atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java

```diff
@@ -189,29 +191,28 @@ public Set<CassandraServer> refreshTokenRangesAndGetServers() {
 
             // Attempt to re-resolve addresses from the configuration; this is important owing to certain race
             // conditions where the entire pool becomes invalid between refreshes.
-            Set<CassandraServer> resolvedConfigAddresses = getCurrentServerListFromConfig();
+            ImmutableSet<CassandraServer> resolvedConfigAddresses = getCurrentServerListFromConfig();
 
-            Set<CassandraServer> lastKnownAddresses = tokenMap.asMapOfRanges().values().stream()
+            ImmutableSet<CassandraServer> lastKnownAddresses = tokenMap.asMapOfRanges().values().stream()
                     .flatMap(Collection::stream)
-                    .collect(Collectors.toSet());
+                    .collect(ImmutableSet.toImmutableSet());
 
-            return Sets.union(resolvedConfigAddresses, lastKnownAddresses);
+            return Sets.union(resolvedConfigAddresses, lastKnownAddresses).immutableCopy();
         }
     }
 
     /**
      * It is expected that config provides list of servers that are directly reachable and do not require special IP
      * resolution.
      * */
-    public Set<CassandraServer> getCurrentServerListFromConfig() {
-        Set<InetSocketAddress> inetSocketAddresses = getServersSocketAddressesFromConfig();
-        return inetSocketAddresses.stream()
+    public ImmutableSet<CassandraServer> getCurrentServerListFromConfig() {
+        return getServersSocketAddressesFromConfig().stream()
                 .map(cassandraHost -> CassandraServer.of(cassandraHost.getHostString(), cassandraHost))
-                .collect(Collectors.toSet());
+                .collect(ImmutableSet.toImmutableSet());
     }
 
-    private Set<InetSocketAddress> getServersSocketAddressesFromConfig() {
-        return runtimeConfig.get().servers().accept(new ThriftHostsExtractingVisitor());
+    private ImmutableSet<InetSocketAddress> getServersSocketAddressesFromConfig() {
+        return ImmutableSet.copyOf(runtimeConfig.get().servers().accept(ThriftHostsExtractingVisitor.INSTANCE));
```

Comment:
I see that `ImmutableSet.copyOf` won't necessarily copy the elements if the underlying set is already immutable (which it will be in this instance, as the visitor is extracting elements from an `Immutables` object, which uses `ImmutableSet` under the hood). This does still do a cast and `instanceof` check; ideally we'd be able to change that so the return type is `ImmutableSet`, but this is probably still acceptable as-is.

Reply:
Similar to comment above, this change was getting quite large without refactoring public method return types, but happy to adjust those as well